db_context/database.py

```
@@ -686,45 +703,112 @@ async def search_in_database(self, search_term: str, limit: int = 20) -> List[st
+                return {"error": "Read-only mode: only SELECT statements are permitted."}
+
+            if self.thick_mode:
+                cursor.execute(sql, params or {})
+            else:
+                await cursor.execute(sql, params or {})
+
+            # Check if this is a SELECT query (has description)
+            if cursor.description:
+                if self.thick_mode:
+                    rows = cursor.fetchmany(max_rows)  # type: ignore[misc]
+                else:
+                    rows = await cursor.fetchmany(max_rows)  # type: ignore[misc]
+                
+                rows = list(rows)  # type: ignore[arg-type]
+                columns = [desc[0] for desc in cursor.description] if cursor.description else []
+                result_rows = [dict(zip(columns, row)) for row in rows]
+                
+                return {
+                    "columns": columns,
+                    "rows": result_rows,
+                    "row_count": len(result_rows)
+                }
+            else:
+                row_count = cursor.rowcount
+                await self._commit(conn)
+                return {
+                    "columns": [],
+                    "rows": [],
+                    "row_count": row_count,
+                    "message": f"Statement executed successfully. {row_count} row(s) affected."
+                }
             
+        except oracledb.Error as e:
+            return {"error": str(e)}
+        except PermissionError as e:
+            return {"error": str(e)}
         finally:
             await self._close_connection(conn)
-    
+
     async def explain_query_plan(self, query: str) -> Dict[str, Any]:
-        """Get execution plan for a SQL query"""
+        """
+        Get the execution plan for a given SQL query and provide optimization suggestions.
+        This tool uses 'EXPLAIN PLAN FOR' to analyze the query without executing it.
+        """
+        if self.read_only:
+            return {
```

We should probably raise an error and handle on client side? Returning a formatted JSON object seems to be mixing concerns.

---

For consistency, use raise PermissionError("Read-only mode: write operations are disabled") by calling the assert_write_allowed. This should be the only right way to check for write permissions accross the database.py module.